ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c

```diff
@@ -37,45 +39,130 @@
 void init_filter_nnfw (void) __attribute__ ((constructor));
 void fini_filter_nnfw (void) __attribute__ ((destructor));
 
+typedef struct
+{
+  nnfw_tensorinfo i_in;
+  nnfw_tensorinfo i_out;
+  nnfw_session *session;
+  gchar *model_path;
+} nnfw_pdata;
+
+static void nnfw_close (const GstTensorFilterProperties * prop,
+    void **private_data);
+
 /**
  * @brief The standard tensor_filter callback
  */
-static int nnfw_open (const GstTensorFilterProperties * prop,
-    void **private_data)
+static int
+nnfw_open (const GstTensorFilterProperties * prop, void **private_data)
 {
-  return 0;
+  NNFW_STATUS status;
+  int err = 0;
+  nnfw_pdata *pdata;
+
+  if (*private_data != NULL) {
+    pdata = *private_data;
+    if (strcmp (prop->model_file, pdata->model_path)) {
+      nnfw_close (prop, private_data);  /* "reopen" */
+    } else {
+      return 1;
+    }
+  }
+
+  pdata = g_new0 (nnfw_pdata, 1);
+  if (pdata == NULL)
+    return -ENOMEM;
+
+  *private_data = (void *) pdata;
+
+  status = nnfw_create_session (&pdata->session);
+  if (status != NNFW_STATUS_NO_ERROR) {
+    err = -EINVAL;
+    g_printerr ("Cannot create nnfw-runtime session");
+    goto unalloc_exit;
+  }
+
+  status = nnfw_load_model_from_file (pdata->session, prop->model_file);
+  if (status != NNFW_STATUS_NO_ERROR) {
+    err = -EINVAL;
+    g_printerr ("Cannot load the model file: %s", prop->model_file);
+    goto unalloc_exit;
+  }
+
+  status = nnfw_prepare (pdata->session);
+  if (status != NNFW_STATUS_NO_ERROR) {
+    err = -EINVAL;
+    g_printerr ("nnfw-runtime cannot prepare the session for %s",
+        prop->model_file);
+    goto unalloc_exit;
+  }
+
+  pdata->model_path = g_strdup (prop->model_file);
+
+  goto exit;
+
+unalloc_exit:
+  g_free (pdata);
+  *private_data = NULL;
+exit:
+  return err;
 }
 
 /**
  * @brief The standard tensor_filter callback
+ * @todo Determine if we need to do "assert" for close-failure.
  */
-static void nnfw_close (const GstTensorFilterProperties * prop,
-    void **private_data)
+static void
+nnfw_close (const GstTensorFilterProperties * prop, void **private_data)
 {
+  nnfw_pdata *pdata;
+  pdata = *private_data;
+
+  if (pdata && pdata->session) {
+    NNFW_STATUS status = nnfw_close_session (pdata->session);
+
+    if (status != NNFW_STATUS_NO_ERROR) {
+      g_printerr ("cannot close nnfw-runtime session for %s",
+          pdata->model_path);
+    }
+  } else {
+    g_printerr ("nnfw_close called without proper nnfw_open");
+    return;
```

Comment:
free model_path and pdata
when called close callback in tensor_filter, private_data will be set null.

Reply:
Updated.